source/postprocess/visualization/named_additional_outputs.cc

```diff
@@ -0,0 +1,172 @@
+                                                   this->n_compositional_fields());
+        MaterialModel::MaterialModelOutputs<dim> out(n_quadrature_points,
+                                                     this->n_compositional_fields());
+
+        in.position = input_data.evaluation_points;
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            Tensor<2,dim> grad_u;
+            for (unsigned int d=0; d<dim; ++d)
+              {
+                grad_u[d] = input_data.solution_gradients[q][d];
+                in.velocity[q][d] = input_data.solution_values[q][this->introspection().component_indices.velocities[d]];
+                in.pressure_gradient[q][d] = input_data.solution_gradients[q][this->introspection().component_indices.pressure][d];
+              }
+
+            in.strain_rate[q] = symmetrize (grad_u);
+
+            in.pressure[q] = input_data.solution_values[q][this->introspection().component_indices.pressure];
+            in.temperature[q] = input_data.solution_values[q][this->introspection().component_indices.temperature];
+
+            for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+              in.composition[q][c] = input_data.solution_values[q][this->introspection().component_indices.compositional_fields[c]];
+          }
+
+        this->get_material_model().create_additional_named_outputs(out);
+        this->get_material_model().evaluate(in, out);
+
+        for (unsigned int k=0; k<out.additional_outputs.size(); ++k)
+          {
+            const MaterialModel::NamedAdditionalMaterialOutputs<dim> *result
+              = dynamic_cast<const MaterialModel::NamedAdditionalMaterialOutputs<dim> *> (out.additional_outputs[k].get());
+
+            if (result)
+              {
+                std::vector<double> outputs(n_quadrature_points);
+                for (unsigned int i=0; i<get_names().size(); ++i)
+                  {
+                    outputs = result->get_nth_output(i);
+
+                    for (unsigned int q=0; q<n_quadrature_points; ++q)
+                      computed_quantities[q][i] = outputs[q];
+                  }
+              }
+          }
+      }
+    }
+  }
+}
+
+
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(NamedAdditionalOutputs,
+                                                  "named additional outputs",
+                                                  "A visualization output object that generates output "
+                                                  "for all additional material outputs that are derived "
+                                                  "from the 'NamedAdditionalMaterialOutputs' base class "
+                                                  "and filled by the material model in the evaluate function.")
```

Comment:
I don't think a typical user would understand this. Maybe start with the use case and write something of the sort: "Some material models can compute quantities other than those that typically appear in the equations that \\aspect{} solves (such as the viscosity, density, etc). Examples of quantities material models may be able to compute are seismic velocities, or other quantities that can be derived from the state variables and the material coefficients such as the stress or stress anisotropies. These quantities are generically referred to as `named outputs' because they are given an explicit name different from the usual outputs of material models.\n\n
This visualization postprocessor outputs whatever quantities the material model can compute. What quantities these are is specific to the material model in use for a simulation, and for many models in fact does not contain any named outputs at all."